include/xtensor/xfunction.hpp

```diff
@@ -374,9 +411,6 @@ namespace xt
 
         tuple_type m_e;
         functor_type m_f;
-        mutable inner_shape_type m_shape;
-        mutable bool m_shape_trivial;
-        mutable bool m_shape_computed;
```

Comment:
```cpp
using cache_type = xfunction_cache<detail::promote_index<typename std::decay_t<CT>::shape_type...>;
mutable cache_type m_cache;
```

Thus later you can replace `this->m_shape` with `m_cache.m_shape` which is a bit more meaningful.